packages/flutter/lib/src/widgets/editable_text.dart

```diff
@@ -263,6 +269,50 @@ class TextEditingController extends ValueNotifier<TextEditingValue> {
   bool _isSelectionWithinComposingRange(TextSelection selection) {
     return selection.start >= value.composing.start && selection.end <= value.composing.end;
   }
+
+  List<TextInputFormatter>? _inputFormatters;
+  void _setInputFormatters(List<TextInputFormatter> newValue) {
+    // The setter does not take null values: if currentValue is null that means
+    // this is the first formatter list ever set, and we should not reformat.
+    final List<TextInputFormatter>? currentValue = _inputFormatters;
+    _inputFormatters = newValue;
+    if (newValue == currentValue || currentValue == null) {
+      return;
+    }
+
+    final Iterator<TextInputFormatter> oldFormatters = currentValue.iterator;
+    final Iterator<TextInputFormatter> newFormatters = newValue.iterator;
+
+    bool needsReformat = false;
+    // Zip the new and the old input formatters lists to determine how many
+    // skippable formatters there are.
+    while (oldFormatters.moveNext() && newFormatters.moveNext()) {
+      if (newFormatters.current.shouldReformat(oldFormatters.current)) {
+        needsReformat = true;
+        break;
+      }
+    }
+
+    TextEditingValue formatted = value;
+    // If we are replacing formatter f(g(x)) with f'(g'(x)), and g' does not need
+    // reformat from g, then a necessary condition that we can skip applying
+    // g' when switching to f'(g'(x)) is that for all fixed points of g,
+    // applying f still results in a fixed point of g.
```

Comment:
Could you re-explain this paragraph to me?

Reply:
I was thinking since this implementation is based on a sufficient condition maybe the comment can help further maintainers find more optimization opportunities. But I re-did the proof and it seems to be a dead end. 
Removed since it doesn't seem to be helpful at all and it was not quite correct.